src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJCourseFactory.java

```diff
@@ -0,0 +1,33 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.CourseFactory;
+import fi.aalto.cs.apluscourses.model.Module;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+
+public class IntelliJCourseFactory implements CourseFactory {
+
+  @NotNull
+  private final Project project;
+
+  public IntelliJCourseFactory(@NotNull Project project) {
+    this.project = project;
+  }
+
+  @Override
+  public Course createCourse(@NotNull String name,
+                             @NotNull List<Module> modules,
+                             @NotNull Map<String, String> requiredPlugins) {
```

Comment:
to consider: would it make sense to wrap `Map<String, String>` with a custom container class or are we still OK with tossing it around?

Reply:
It certainly would. I didn't however refactor much `Module` class in this PR, so I left it unchanged.